crypto/ppccap.c

```diff
@@ -171,6 +171,37 @@ void ecp_nistz256_from_mont(unsigned long res[4], const unsigned long in[4])
 }
 #endif
 
+size_t SHA3_absorb(uint64_t A[5][5], const unsigned char *inp, size_t len,
+    size_t r);
+void SHA3_squeeze(uint64_t A[5][5], unsigned char *out, size_t len, size_t r);
+
+size_t SHA3_absorb_default(uint64_t A[5][5], const unsigned char *inp,
+    size_t len, size_t r);
+void SHA3_squeeze_default(uint64_t A[5][5], unsigned char *out, size_t len,
+    size_t r);
+
+size_t SHA3_absorb_vsx(uint64_t A[5][5], const unsigned char *inp,
+    size_t len, size_t r);
+void SHA3_squeeze_vsx(uint64_t A[5][5], unsigned char *out, size_t len,
+    size_t r);
+
+size_t SHA3_absorb(uint64_t A[5][5], const unsigned char *inp, size_t len,
+    size_t r)
+{
+    return OPENSSL_ppccap_P & PPC_CRYPTO207
+        ? SHA3_absorb_vsx(A, inp, len, r)
+        : SHA3_absorb_default(A, inp, len, r);
+}
+
+void SHA3_squeeze(uint64_t A[5][5], unsigned char *out, size_t len, size_t r)
```

Comment:
Where would you like me to put it? Is there a particular order you want the algorithms?

Reply:
Looks like this code is better placed in the cypto/sha/keccak1600.c file

Reply:
I am happy to put it there, but this code calls the _vsx variants of the code if the processor has support for them, which are powerpc specific. If you look above in ppccap.c, the same is done for chacha, poly1305, and ECP_NISTZ256_ASM. I can move it, but I was just following suit. Please confirm.

Reply:
Should I change it or leave it?

Reply:
Looking at the other *cap.c files none of them (with one exception) seem to have algorithm specific code in them. The one exception is sparcv9cap.c which does have a BN function defined in it. This looks similarly misplaced to me.

I think the code should be moved out of this file.

Reply:
I am looking at:
https://github.com/openssl/openssl/blob/master/crypto/ppccap.c#L105 (chacha)
and
https://github.com/openssl/openssl/blob/master/crypto/ppccap.c#L132 (poly)

Which are doing the same kind of function selection I am doing. I am I missing something?

Reply:
I can understand your thinking here. IMO this is still misplaced - and the other algorithm specific code in this file is also misplaced. However, it is perhaps a little much to ask you to resolve this more general issue in this PR. So I'm ok with it being left here for now. It would be nice at some point in the future if someone could properly refactor this and put things in the right place - but that doesn't have to be done now.

Reply:
Thanks @mattcaswell , but I am going to undo the whole thing since @dot-asm wrote below that the performance impact is small and we are better off turning off the assembly for powerpc.